### ISSUE TITLE
Adding device/sensor name to get-sensorhistory

### ISSUE THREAD

**Issue by SamuelmAsh** [lordmilko/PrtgAPI#26]
Hi LordMilko.

I am extracting lists of all traffic sensors so i can then filter in excel those which dont ever pass any data so i can send it over to networks to get them removed. Im using get-sensor history for this however i cannot work out how to pass device/probe names to the outputting CSV. 

I have tried -Select in various places. Any guidance would be really appreciated. 

Kindest
Samuel

**Comment by lordmilko** [comment]
Hi Samuel,

You can either use `Select-Object` and specify a [custom expression](https://4sysops.com/archives/add-a-calculated-property-with-select-object-in-powershell/) (which is messy) or append the properties you're after with the `Add-Member` cmdlet (messy with multiple members)

```powershell
Get-Sensor -count 1 -Status Up | foreach { $_ | Get-SensorHistory | Add-Member Device $_.Device -PassThru | Add-Member Probe $_.Probe -PassThru } | Select *
```

Note that without piping the result of `Get-SensorHistory` to `Select-Object` and specifying to retrieve the newly added properties (or * for all properties) you wouldn't *see* the newly added properties when piping to `Format-List` as `Get-SensorHistory` has custom Table/List views that don't include these newly added members. When exported to a CSV however, all members that exist will be included

Regards,
lordmilko

**Comment by lordmilko** [comment]
As you mentioned you are doing some stuff with traffic sensors, please note there is a bug in the current version of PrtgAPI (0.8.6) wherein `Get-SensorHistory` does not show both the *speed* and *volume* channels for traffic sensors. It only shows the *volume channel*. If you need to operate on the *speed* channel, you will need to invoke the underlying C# version directly (which can still be done within PowerShell)

```powershell
C:\> (Get-PrtgClient).GetSensorHistory(9519)

DateTime              SensorId ChannelRecords
--------              -------- --------------
20/07/2018 8:35:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:30:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:25:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:20:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:15:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:10:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:05:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:00:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:55:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:50:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:45:00 PM     9519 {Total: 0.14 MByte, Total: < 0.01 Mbit/s, Traff...
20/07/2018 8:40:00 PM     9519 {Total: 0.04 MByte, Total: < 0.01 Mbit/s, Traff...
```

Each `SensorHistoryData` record contains an array of `ChannelHistoryRecord` objects

```powershell
C:\> (Get-PrtgClient).GetSensorHistory(9519)[0].ChannelRecords|ft

DateTime              SensorId Name       ChannelId Value
--------              -------- ----       --------- -----
20/07/2018 8:40:00 PM     9519 Total             -1 0.14 MByte
20/07/2018 8:40:00 PM     9519 Total             -1 < 0.01 Mbit/s
20/07/2018 8:40:00 PM     9519 Trafficin          0 0.08 MByte
20/07/2018 8:40:00 PM     9519 Trafficin          0 < 0.01 Mbit/s
20/07/2018 8:40:00 PM     9519 Trafficout         1 0.06 MByte
20/07/2018 8:40:00 PM     9519 Trafficout         1 < 0.01 Mbit/s
20/07/2018 8:40:00 PM     9519 Packets            2 2,102 #
20/07/2018 8:40:00 PM     9519 Packets            2 7 #/s
20/07/2018 8:40:00 PM     9519 Downtime          -4 0 %
```

You can see that there are two Total, Trafficin, Trafficout and Packets channels. One of these is the volume channel, the other is the speed channel (the order will be the same for every sensor). As PrtgAPI does not label these correctly, they appear as the "same" channel, which is why they aren't displayed correctly in the output from `Get-SensorHistory`. This issue will be resolved in PrtgAPI 0.9.0

In the meantime however, if you need to operate on speed channels you will need to do a bit of manual scripting in order to access them, rather than simply executing one liners

**Comment by SamuelmAsh** [comment]
Hi, thanks for your detailed reply, i dont need speed at this time so powershell will work just fine for my purposes. Thanks for the guide though, i am planning to move to the C# version of the PRTGApi soon :) 

Your PS solution worked perfectly. Thanks a lot for your help
